torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -200,21 +203,39 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
       // a use with a different profiled type, we will still be correct.
       // In the future we could consider unifying the types of uses, or adding a
       // type refinement node so uses can have the correct corresponding type.
-      if (profiled_type == TensorType::get()) {
+      if (profiled_type == TensorType::get() && !seen_none) {
         continue;
       }
+
+      TensorTypePtr input_tensor_type = nullptr;
+      bool input_is_optional = false;
+      if (it->input()->type()->kind() == c10::TypeKind::TensorType) {
+        input_tensor_type = it->input()->type()->expect<TensorType>();
+      } else {
+        input_tensor_type = it->input()
+                                ->type()
+                                ->expectRef<OptionalType>()
+                                .getElementType()
+                                ->expect<TensorType>();
+        input_is_optional = true;
+      }
+
       // If we encounter non-identical profiled types for the same value, merge
       // them.  This situation can happen if, e.g., loop unrolling duplicates
       // profiled types in a loop body in a manner that isn't logically
       // consistent (see TestTEFuser.test_unrolled_cat).
-      auto input_type = it->input()->type()->expect<TensorType>();
-      if (input_type == TensorType::get()) {
-        it->input()->setType(profiled_type);
+      TensorTypePtr merged_tensor_type = profiled_type;
+      if (input_tensor_type != TensorType::get()) {
+        merged_tensor_type = input_tensor_type->merge(*profiled_type);
+      }
+
+      if (seen_none || input_is_optional) {
+        it->input()->setType(OptionalType::create(merged_tensor_type));
       } else {
-        it->input()->setType(input_type->merge(*profiled_type));
+        it->input()->setType(merged_tensor_type);
       }
```

Comment:
@Krovatkin how should this behave?

1) an Optional[Tensor] always gets specialized to `Optional[specialized-tensor]`, e.g. if profiler sees "tensor(2, 2), tensor(2, 3), and tensor(2, 4)", then it'll be `Optional[Float(2, ?)]`
2) Optional[Tensor] can get specialized to `specialized-tensor` if the profiler never sees `None`, or `Optional[specialized-tensor]` if the profiler does see `None`

Reply:
@davidberard98 I'd suggest 1. or to even be on the safe side to always set the type to `Optional::of(Tensor::get)` as if we never even profiled the optional tensor types. I'd like to keep this PR focused on *profiling* optional tensors rather than *extending optimizations to consume this new profiled type* and option 2 seems to require cascading changes in TE and possibly in NVFuser.